File: Java/프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.java
import java.util.*;

class Solution {
    public long solution(int[] weights) {
        HashMap<Integer, Long> hm = new HashMap<>();
        
        for (int weight : weights) {
            hm.put(weight, hm.getOrDefault(weight, (long) 0) + 1);
        }
        
        long answer = 0;
        for (Map.Entry<Integer, Long> entry : hm.entrySet()) {
            int key = entry.getKey();
            long value = entry.getValue();
            
            if (value > 1) {
                answer += (value * (value - 1)) / 2;
            }
            
            if (hm.containsKey(key * 2)) {
                answer += hm.get(key * 2) * value;
            }
            
            if ((key * 3) % 2 == 0 && hm.containsKey((key * 3) / 2)) {
                answer += hm.get((key * 3) / 2) * value;
            }
            
            if ((key * 4) % 3 == 0 && hm.containsKey((key * 4) / 3))  {
                answer += hm.get((key * 4) / 3) * value;
            }
        }
        
        return answer;
    }
}

File: Swift/프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.swift
import Foundation

func solution(_ weights:[Int]) -> Int64 {
    var weightDic: [Int : Int64] = [:]
    for weight in weights {
        weightDic[weight] = (weightDic[weight] ?? 0) + 1
    }

    var answer: Int64 = 0
    for dic in weightDic {
        let key = dic.key
        let value = dic.value
        
        if value > 1 {
            answer += (((value) * (value - 1)) / 2)
        }
        
        if (key * 2) % 3 == 0,let pair = weightDic[(key * 2) / 3] {
            answer += pair * value
        }
        
        if (key * 3) % 4 == 0, let pair = weightDic[(key * 3) / 4] {
            answer += pair * value
        }
        
        if key % 2 == 0, let pair = weightDic[key / 2] {
            answer += pair * value
        }
    }
    
    return answer
}